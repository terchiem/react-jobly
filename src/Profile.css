.Profile {
  width: 40vw;
  margin: 30px auto;
  padding: 10px 20px;
  background-color: #fff;
  border: 1px solid #ccc;
  box-shadow: 1px 1px 2px 0 rgba(0, 0, 0, 0.2);
  position: relative;
}

.Profile h2,
.Profile h3 {
  margin: 0;
}