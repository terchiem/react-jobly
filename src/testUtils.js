import React from 'react';
import UserContext from './UserContext';

const testUser = {
  username: "test",
  first_name: "test",
  last_name: "test",
  email: "test@email.com",
  photo_url: null
}

const testToken = "test";

function UserProvider({ children, currentUser=testUser, token=testToken }) {
  return (
    <UserContext.Provider value={{currentUser, token}}>
      {children}
    </UserContext.Provider>
  )
};

export { UserProvider };