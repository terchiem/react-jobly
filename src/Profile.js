import React, { useContext, useState, useEffect } from 'react';
import UserContext from './UserContext';
import JoblyApi from './JoblyApi';
import './Profile.css';

import LoadingSpinner from './LoadingSpinner';

/** Displays a user's details as a form that allows the user to edit 
 * 
 * Props:
 *  setEditedUser -> function from App to set newly edited user as currentUser
 * 
 * Context:
 *  currentUser -> universal current user object
 * 
 * State:
 *  formData -> state for form inputs
*/

function Profile({ setEditedUser }) {
  const { currentUser } = useContext(UserContext);

  const [formData, setFormData] = useState({
    password: "",
    first_name: "",
    last_name: "",
    email: "",
    photo_url: "",
    errors: []
  });

  const [updateSuccess, setUpdateSuccess] = useState(false);

  // when currentUser data has loaded, set the formData values
  useEffect(() => {
    setFormData({
      password: "",
      first_name: currentUser.first_name || "",
      last_name: currentUser.last_name || "",
      email: currentUser.email || "",
      photo_url: currentUser.photo_url || "",
      errors: []
    });
  }, [currentUser]);

  /** Update state on input change */
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(data => ({...data, [name]: value}));
    setUpdateSuccess(false);
  }

  /** Call api to log in, stores token in local storage and updates state */
  async function editUser(evt) {
    evt.preventDefault();
    
    try {
      const editedUser = await JoblyApi.editUser(currentUser.username, formData);
      setEditedUser(editedUser);
      setUpdateSuccess(true);
    } catch (errors) {
      // fill formData errors array to display
      setFormData(data => ({...data, errors}));
    }
  }

  /** Renders form errors if present */
  function renderErrors() {
    return (formData.errors.length ?
      <ul className="Form-errors">
        {formData.errors.map((message, i) => (
          <li key={i}>{message}</li>
        ))}
      </ul> : null
    )
  }

  if (!currentUser) {
    return <LoadingSpinner />;
  }

  return ( 
    <div className="Profile">
      <form className="Form" onSubmit={editUser}>
        <h2>Edit Profile</h2>
        <div className="Form-group">
          <h3>Username</h3>
          <p>{currentUser.username}</p>
        </div>
        <div className="Form-group">
          <label className="Form-label" htmlFor="first_name">First name</label>
          <input
            className="Form-input" 
            id="first_name" 
            name="first_name" 
            value={formData.first_name} 
            onChange={handleChange} 
          />
        </div>
        <div className="Form-group">
          <label className="Form-label" htmlFor="last_name">Last name</label>
          <input
            className="Form-input" 
            id="last_name" 
            name="last_name" 
            value={formData.last_name} 
            onChange={handleChange} 
          />
        </div>
        <div className="Form-group">
          <label className="Form-label" htmlFor="email">Email</label>
          <input
            className="Form-input" 
            id="email" 
            name="email" 
            type="email"
            value={formData.email} 
            onChange={handleChange} 
          />
        </div>
        <div className="Form-group">
          <label className="Form-label" htmlFor="photo_url">Photo URL</label>
          <input
            className="Form-input" 
            id="photo_url" 
            name="photo_url" 
            value={formData.photo_url} 
            onChange={handleChange} 
          />
        </div>
        <div className="Form-group">
          <label htmlFor="password" className="Form-label">Confirm Password</label>
          <input 
            className="Form-input"
            id="password" 
            name="password" 
            type="password"
            value={formData.password} 
            onChange={handleChange} 
          />
        </div>
        <button className="Form-submit">Submit</button>
  
        { renderErrors() }
        { updateSuccess ? <div className="Form-success">Saved changes!</div> : null }
      </form>
    </div> 
  )
}

export default Profile;