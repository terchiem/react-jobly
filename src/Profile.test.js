import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import Profile from './Profile';
import { UserProvider } from './testUtils';

test('renders without crashing', () => {
  render(
    <MemoryRouter>
      <UserProvider>
        <Profile />
      </UserProvider>
    </MemoryRouter>
  );
});

test('matches snapshot', () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <Profile />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
