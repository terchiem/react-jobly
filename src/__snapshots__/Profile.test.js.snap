// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Profile"
  >
    <form
      class="Form"
    >
      <h2>
        Edit Profile
      </h2>
      <div
        class="Form-group"
      >
        <h3>
          Username
        </h3>
        <p>
          test
        </p>
      </div>
      <div
        class="Form-group"
      >
        <label
          class="Form-label"
          for="first_name"
        >
          First name
        </label>
        <input
          class="Form-input"
          id="first_name"
          name="first_name"
          value="test"
        />
      </div>
      <div
        class="Form-group"
      >
        <label
          class="Form-label"
          for="last_name"
        >
          Last name
        </label>
        <input
          class="Form-input"
          id="last_name"
          name="last_name"
          value="test"
        />
      </div>
      <div
        class="Form-group"
      >
        <label
          class="Form-label"
          for="email"
        >
          Email
        </label>
        <input
          class="Form-input"
          id="email"
          name="email"
          type="email"
          value="test@email.com"
        />
      </div>
      <div
        class="Form-group"
      >
        <label
          class="Form-label"
          for="photo_url"
        >
          Photo URL
        </label>
        <input
          class="Form-input"
          id="photo_url"
          name="photo_url"
          value=""
        />
      </div>
      <div
        class="Form-group"
      >
        <label
          class="Form-label"
          for="password"
        >
          Confirm Password
        </label>
        <input
          class="Form-input"
          id="password"
          name="password"
          type="password"
          value=""
        />
      </div>
      <button
        class="Form-submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
